unlock.go

```diff
@@ -143,24 +165,91 @@ func UnlocksSigUniqueAndRefValidator(api API) UnlockValidatorFunc {
 				return ierrors.Wrapf(ErrSigUnlockNotUnique, "signature unlock block at index %d is the same as %d", index, existingIndex)
 			}
 
-			seenSigUnlockBytes[string(sigBlockBytes)] = index
 			seenSigUnlocks[uint16(index)] = struct{}{}
+			seenSigUnlockBytes[string(sigBlockBytes)] = index
+
 		case ReferentialUnlock:
-			if prevRef := seenRefUnlocks[x.Ref()]; prevRef != nil {
-				if !x.Chainable() {
-					return ierrors.Wrapf(ErrReferentialUnlockInvalid, "%d references existing referential unlock %d but it does not support chaining", index, x.Ref())
+			if prevRef := seenRefUnlocks[unlock.Ref()]; prevRef != nil {
+				if !unlock.Chainable() {
+					return ierrors.Wrapf(ErrReferentialUnlockInvalid, "%d references existing referential unlock %d but it does not support chaining", index, unlock.Ref())
 				}
-				seenRefUnlocks[uint16(index)] = x
+				seenRefUnlocks[uint16(index)] = unlock
 
 				break
 			}
-			// must reference a sig unlock here
-			if _, has := seenSigUnlocks[x.Ref()]; !has {
-				return ierrors.Wrapf(ErrReferentialUnlockInvalid, "%d references non existent unlock %d", index, x.Ref())
+
+			// must reference a sig or multi unlock here
+			_, hasSigUnlock := seenSigUnlocks[unlock.Ref()]
+			_, hasMultiUnlock := seenMultiUnlocks[unlock.Ref()]
+			if !hasSigUnlock && !hasMultiUnlock {
+				return ierrors.Wrapf(ErrReferentialUnlockInvalid, "%d references non existent unlock %d", index, unlock.Ref())
+			}
+			seenRefUnlocks[uint16(index)] = unlock
+
+		case *MultiUnlock:
+			multiUnlockBytes, err := api.Encode(unlock)
+			if err != nil {
+				return ierrors.Errorf("unable to serialize multi unlock block at index %d for dup check: %w", index, err)
+			}
+
+			if existingIndex, exists := seenMultiUnlockBytes[string(multiUnlockBytes)]; exists {
+				return ierrors.Wrapf(ErrMultiUnlockNotUnique, "multi unlock block at index %d is the same as %d", index, existingIndex)
+			}
+
+			for subIndex, subU := range unlock.Unlocks {
+				switch subUnlock := subU.(type) {
+				case *SignatureUnlock:
+					if subUnlock.Signature == nil {
+						return ierrors.Wrapf(ErrSigUnlockHasNilSig, "at index %d.%d is nil", index, subIndex)
+					}
+
+					sigBlockBytes, err := api.Encode(subUnlock.Signature)
+					if err != nil {
+						return ierrors.Errorf("unable to serialize signature unlock block at index %d.%d for dup check: %w", index, subIndex, err)
+					}
+
+					if existingIndex, exists := seenSigUnlockBytes[string(sigBlockBytes)]; exists {
+						return ierrors.Wrapf(ErrSigUnlockNotUnique, "signature unlock block at index %d.%d is the same as %d", index, subIndex, existingIndex)
```

Comment:
This is a problem because I can have restricted addresses with the same pub key and just different capabilities bits set. This means they will have the exact same signature. The uniqueness checks in the array rules do not account for this.

Reply:
This was fixed by removing all existing restricted address types, except the multi address.
Now the uniqueness of addresses/pubkeys inside the multi unlock should be enforced again.

Reply:
I re-introduced the RestrictedEd25519 address again, so I had to solve this in a different way.
I added a `PubKeyHash` function to the address interface and added a custom validator that checks the uniqueness of these pubKeyHashes in the Address slice.

Reply:
Reverted the `PublicKeyHash` function changes. It's now a type switch in the validation.